### ISSUE TITLE
Support Redeploying an Index

### ISSUE THREAD

**Issue by danshultz** [ember-cli/ember-cli-deploy#105]
Currently, ember cli deploy prevents a duplicate index from being deployed in order to protect someone from accidentally re-deploying an already deployed index. It would be helpful for ember cli deploy to allow the redeployment of an already deployed index. Most likely a `--force` option would be best however, maybe a `deploy:purge --revision` option could also be reasonable.

This issue is to open the discussion and help define a solution to let a user opt into redeploying an index if needed.

**Comment by achambers** [comment]
I too have come across this need, however it has only been during development.  This is probably due to the fact that anything outside of that is deployed by a CI server, and therefore, a new git hash is present as this is how the CI server is kicked off.

It would be interesting to understand the use case here as I'm sure there are people still deploying from their local machines which I suspect is where this need arises.

Would it makes sense to allow the `--force` or `deploy:purge` options to only be available for the `development` env?

Forcing could be somewhat dangerous.  But then, so can `git push -f`.

Anyway, as mentioned above, I have felt the pain of this though, however it has only been during development in which case I can squish any of those unwanted commits before merging into master.

**Comment by lukemelia** [comment]
I've also felt some pain from this during initial configuration of ember-cli-deploy on a project. I'm +1 allowing override.

**Comment by justinko** [comment]
+1 --- forgot to add the fingerprint prepend!

**Comment by achambers** [comment]
Yep, I've started to come across it as well. My vote is probably for
`--force` option

**Comment by mrandre** [comment]
+1

**Comment by lukemelia** [comment]
I've recently taken a new approach to this. We can solve this better, but in the interest of stimulating ideas:

```js
// Brocfile.js
var crypto = require('crypto');

var EmberApp = require('ember-cli/lib/broccoli/ember-app');

function md5Hash(buf) {
  var md5 = crypto.createHash('md5');
  md5.update(buf);
  return md5.digest('hex');
}
var FINGERPRINT_VERSION = '1'; // increment this to bust cache on s3/cloudfront

var app = new EmberApp({
  fingerprint: {
    prepend: 'https://' + process.env.CDN_ASSET_HOST + '/',
    customHash: function(buf){
      return md5Hash(buf) + FINGERPRINT_VERSION;
    }
  }
});
```

By incrementing the FINGERPRINT_VERSION constant, you can be guaranteed new filenames without the original contents having changed.

**Comment by thehydroimpulse** [comment]
This is definitely needed for other tagging systems, like using a git branch or an arbitrary label.

**Comment by LevelbossMike** [comment]
@thehydroimpulse You can deploy commits on feature branch as they get a unique sha.

**Comment by thehydroimpulse** [comment]
@LevelbossMike I meant storing the actual branch as the version tag (instead of storing the sha at all) in Redis and the like. To where you could access it via `foobar.com?rev=easy_to_remember`. It's pretty tedious to have to distribute the new sha (everytime you make changes) to coworkers or customers that might want to try beta features. (I think Skylight does this)

**Comment by LevelbossMike** [comment]
ah ok. yeah makes sense.

**Comment by juggy** [comment]
Connecting manually to redis and deleting the key would do it no? 

Then again, an easier way built into cli-deploy would be good.

**Comment by jbescoyez** [comment]
A PR has already been made for this issue: https://github.com/ember-cli/ember-cli-deploy/pull/39 that should be coupled with https://github.com/LevelbossMike/ember-deploy-redis/pull/1 which is still open.

I'm ready to update it if @LevelbossMike and @lukemelia are in :+1: for this implementation.

**Comment by LevelbossMike** [comment]
The redis plugin will be a different one (much improved) for the pipeline implementation of ember-cli-deploy (v0.5.0). The new plugin already supports overriding the currently set index and will be available in the near future with the next release of ember-cli-deploy.

**Comment by jbescoyez** [comment]
@LevelbossMike Awesome!

**Comment by jbescoyez** [comment]
@LevelbossMike I've seen the allowOverwrite option here https://github.com/zapnito/ember-cli-deploy-redis/blob/master/lib/redis.js . However, I do not see it implemented yet in https://github.com/zapnito/ember-cli-deploy-redis/blob/master/lib/redis.js. Am I missing something? If not, I'll do a PR soon.

**Comment by achambers** [comment]
@jbescoyez If you pass `allowOverwrite` into your `redis: {}` config in `deploy.js` then this will be passed through to the plugin.

**Comment by achambers** [comment]
@jbescoyez Wondering whether your last comment has an incorrect link seeing you linked to the same file twice.

**Comment by jbescoyez** [comment]
@achambers Oops. Updated. However, I think you answered my question. I gonna give it a try today. :) Tnx

**Comment by achambers** [comment]
@jbescoyez The file `lib/tasks/deploy-index.js` is not used in the new deploy pipeline.  This will be one of the files that will be deleted in 0.5.0

**Comment by jbescoyez** [comment]
@achambers Ok. However, it is still present in the 0.5.0-dev branch. Is there a branch of `ember-cli-deploy` that is already compatible with `zapnito/ember-cli-deploy-redis` ?

**Comment by achambers** [comment]
@jbescoyez `0.5.0-dev` **is** the branch of ember-cli-deploy that is compatible with [zapnito/ember-cli-deploy-redis](https://github.com/zapnito/ember-cli-deploy-redis)

**Comment by devinus** [comment]
Guys, I'm a little lost on the outcome of this. I'm trying to support a workflow where we can manually trigger deploys though our CI solution, but when ember-cli-deploy wont allow reuploads, I'm a bit stuck.

**Comment by lukemelia** [comment]
@devinus in 0.5.0, the redis plugin supports `allowOverwrite: true`: https://github.com/ember-cli-deploy/ember-cli-deploy-redis#allowoverwrite

**Comment by lukemelia** [comment]
Fixed in 0.5.0, which has in beta now and will be released soon.
